runtimebp/internal/maxprocs/maxprocs.go

```diff
@@ -0,0 +1,142 @@
+	// invalid). It defaults to some "oversubscription" to take advantage of
+	// excess Kubernetes node capacity when available.
+	//
+	// Extremely performance-sensitive services should set scale to 1.0 to
+	// ensure consistent performance.
+	//
+	// This value was NOT chosen scientifically and IS subject to change.
+	defaultCPURequestScale = 1.5
+
+	setByGOMAXPROCS   = "gomaxprocs"
+	setByRequest      = "cpu_request"
+	setByAutomaxprocs = "automaxprocs"
+)
+
+var (
+	mEnvGOMAXPROCS = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "baseplate_go_env_gomaxprocs",
+		Help: "Value of the GOMAXPROCS environment variable at startup. 0 if not a number",
+	}, []string{"present"})
+	mEnvCPURequest = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "baseplate_go_env_baseplate_cpu_request",
+		Help: "Value of the BASEPLATE_CPU_REQUEST environment variable at startup. 0 if not a number",
+	}, []string{"present"})
+	mEnvCPURequestScale = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "baseplate_go_env_baseplate_cpu_request_scale",
+		Help: "Value of the BASEPLATE_CPU_REQUEST_SCALE environment variable at startup. 0 if not a number",
+	}, []string{"present"})
+
+	initialGOMAXPROCS = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "baseplate_go_initial_gomaxprocs",
+		Help: "Resolved value of GOMAXPROCS at startup",
+	}, []string{"set_by"})
+
+	_ = promauto.NewGaugeFunc(prometheus.GaugeOpts{
+		Name: "baseplate_go_current_gomaxprocs",
+		Help: "Current value of GOMAXPROCS",
+	}, currentGOMAXPROCS)
+
+	// overridable for tests
+	setWithAutomaxprocs = func() {
+		ubermaxprocs.Set(
+			ubermaxprocs.Min(2),
+			ubermaxprocs.Logger(func(s string, i ...interface{}) {
+				fmt.Fprintf(os.Stderr, s, i...) // contains context
+			}),
+		)
+	}
+)
+
+// Set configures the runtime's GOMAXPROCS using the following heuristic:
+//   1. If $GOMAXPROCS is set, Set relinquishes control to the Go runtime.
+//      This should cause the runtime to respect this value directly.
+//   2. If $BASEPLATE_CPU_REQUEST is unset/invalid, Set relinquishes control to automaxprocs, minimum 2.
+//      See https://pkg.go.dev/go.uber.org/automaxprocs for specific behavior.
+//   3. Otherwise, $BASEPLATE_CPU_REQUEST is multiplied by $BASEPLATE_CPU_REQUEST_SCALE
+//      (or defaultCPURequestScale) to compute the new GOMAXPROCS, minimum 2.
+//
+// Set also exports several metrics to facilitate further tuning/analysis.
+func Set() {
+	envGOMAXPROCS := &floatEnv{key: "GOMAXPROCS", gauge: mEnvGOMAXPROCS}
+	envCPURequest := &floatEnv{key: "BASEPLATE_CPU_REQUEST", gauge: mEnvCPURequest}
+	envCPURequestScale := &floatEnv{key: "BASEPLATE_CPU_REQUEST_SCALE", gauge: mEnvCPURequestScale}
+
+	for _, env := range []*floatEnv{envGOMAXPROCS, envCPURequest, envCPURequestScale} {
+		env.raw, env.present = os.LookupEnv(env.key)
+		env.val, _ = strconv.ParseFloat(env.raw, 64) // error not needed as we check val > 0 and save raw
```

Comment:
🔕 the comment is not always true? the `>0` check is not handled in this for loop. outside of this for loop, there's <=0 check for `envCPURequest` and >0 check for `envCPURequestScale`, but I don't see any check for `envGOMAXPROCS`?

how about this: we change `present` to 3 states: true, not_a_number (or maybe not_valid), false?